### PR TITLE
Added search request

### PR DESCRIPTION
**Pull request by DZarudnyi**

### CODE REVIEW

**Comment by DZarudnyi** [comment]
So i was wondering how could i add search by part of the field, not exact match?

File: src/main/java/com/example/books/repository/book/BookSpecificationProviderManager.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+@Component
+public class BookSpecificationProviderManager implements SpecificationProviderManager<Book> {
+    private final List<SpecificationProvider<Book>> bookSpecificationProviders;
+
+    @Override
+    public SpecificationProvider<Book> getSpecificationProvider(String key) {
+        return bookSpecificationProviders.stream()
+                .filter(p -> p.getKey().equals(key))
+                .findFirst()
+                .orElseThrow(() -> new RuntimeException(
```
Comment (okuzan) [inline src/main/java/com/example/books/repository/book/BookSpecificationProviderManager.java:20]: let's throw NoSuchElement

File: src/main/java/com/example/books/repository/book/BookSpecificationBuilder.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+public class BookSpecificationBuilder implements SpecificationBuilder<Book> {
+    private final SpecificationProviderManager<Book> bookSpecificationProviderManager;
+
+    @Override
+    public Specification<Book> build(BookSearchParametersDto searchParameters) {
+        Specification<Book> spec = Specification.where(null);
+
+        if (searchParameters.authors() != null && searchParameters.authors().length > 0) {
+            spec = spec.and(bookSpecificationProviderManager.getSpecificationProvider("author")
+                    .getSpecification(searchParameters.authors()));
```
Comment (okuzan) [inline src/main/java/com/example/books/repository/book/BookSpecificationBuilder.java:22]: extract literals to constants

File: src/main/java/com/example/books/repository/book/BookSpecificationProviderManager.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+@RequiredArgsConstructor
+@Component
+public class BookSpecificationProviderManager implements SpecificationProviderManager<Book> {
+    private final List<SpecificationProvider<Book>> bookSpecificationProviders;
+
+    @Override
+    public SpecificationProvider<Book> getSpecificationProvider(String key) {
+        return bookSpecificationProviders.stream()
+                .filter(p -> p.getKey().equals(key))
```
Comment (okuzan) [inline src/main/java/com/example/books/repository/book/BookSpecificationProviderManager.java:18]: p is a bad naming

**Review by okuzan — CHANGES_REQUESTED** [review]

**Review by Ivan95kos — APPROVED** [review]
Nice

**Pull request merged**